pkg/compact/compact.go

```diff
@@ -470,6 +483,200 @@ func (cg *Group) Resolution() int64 {
+	// updating the exposed metrics inside the above loop will change based on iterations needed for each plan loop
+	// updating the metrics' maps directly - some keys may not be present in the groups map; also saves the cost of a lookup if done directly
+	for key, iters := range groupCompactions {
+		ps.ProgressMetrics.NumberOfCompactionRuns.WithLabelValues(key).Add(float64(iters))
+		ps.ProgressMetrics.NumberOfCompactionBlocks.WithLabelValues(key).Add(float64(groupBlocks[key]))
+	}
+
+	return nil
+}
+
+type DownsampleMetrics struct {
+	BlocksDownsampled *prometheus.GaugeVec
+	// - number of blocks to be finally downsampled, grouped by groupKey
+}
+
+type DefaultDownsampleSim struct {
+	*DownsampleMetrics
+}
+
+func NewDefaultDownsampleSim(reg prometheus.Registerer) *DefaultDownsampleSim {
+	return &DefaultDownsampleSim{
+		DownsampleMetrics: &DownsampleMetrics{
+			BlocksDownsampled: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_blocks_downsampled",
+				Help: "number of blocks to be downsampled",
+			}, []string{"group"}),
+		},
+	}
+}
+
+func (ds *DefaultDownsampleSim) ProgressCalculate(ctx context.Context, groups []*Group) error {
+
+	var metas map[ulid.ULID]*metadata.Meta // pre allocate with size
+	// mapping ULIDs to meta - reference: https://github.com/thanos-io/thanos/blob/18049504408c5ae09deb76961b92baf7f96b0e93/pkg/compact/compact.go#L425-L436
+	for _, group := range groups {
+		for _, meta := range group.metasByMinTime {
+			metas[meta.ULID] = meta
+		}
+	}
+
+	sources5m := map[ulid.ULID]struct{}{}
+	sources1h := map[ulid.ULID]struct{}{}
+	metasULIDS := make([]ulid.ULID, 0, len(metas)) // change size of map allocation
+
+	for k, m := range metas {
+		if m.Thanos.Downsample.Resolution != downsample.ResLevel2 {
+			metasULIDS = append(metasULIDS, k)
+		}
+		switch m.Thanos.Downsample.Resolution {
+		case downsample.ResLevel0:
+			continue
+		case downsample.ResLevel1:
+			for _, id := range m.Compaction.Sources {
+				sources5m[id] = struct{}{}
+			}
+		case downsample.ResLevel2:
+			for _, id := range m.Compaction.Sources {
+				sources1h[id] = struct{}{}
+			}
+		default:
+			return errors.Errorf("unexpected downsampling resolution %d", m.Thanos.Downsample.Resolution)
+		}
+	}
+
+	sort.Slice(metasULIDS, func(i, j int) bool {
+		return metasULIDS[i].Compare(metasULIDS[j]) < 0
+	})
+
+	// each of these metas is added to the channel and hence, each of these ULIDs should be processed similar to processDownsampling()
+	for _, mk := range metasULIDS {
+		m := metas[mk]
+
+		switch m.Thanos.Downsample.Resolution {
+		// removed case with ResLevel2 since those aren't included in metasULIDs
+		case downsample.ResLevel0:
+			missing := false
+			for _, id := range m.Compaction.Sources {
+				if _, ok := sources5m[id]; !ok {
+					missing = true
+					break
+				}
+			}
+			if !missing {
+				continue
+			}
+
+			if m.MaxTime-m.MinTime < downsample.DownsampleRange0 {
+				continue
+			}
+
```

Comment:
Update the temp variable for metrics here

Reply:
Done this locally. I'll show it to you today.